lib/Sema/TypeCheckGeneric.cpp

```diff
@@ -78,6 +78,23 @@ void checkGenericParamList(TypeChecker &tc,
     RequirementRequest::visitRequirements(owner, resolution.getStage(),
         [&](const Requirement &req, RequirementRepr *reqRepr) {
           auto source = FloatingRequirementSource::forExplicit(reqRepr);
+          
+          // If we're extending a protocol and adding a redundant requirement,
+          // for example, `extension Foo where Self: Foo`, then emit a
+          // diagnostic.
+          
+          if (auto extDecl = dyn_cast<ExtensionDecl>(owner.dc->getAsDecl())) {
+            auto ownerType = extDecl->getExtendedType();
+            auto selfType = req.getFirstType();
+            auto reqType = req.getSecondType();
+            
+            if (ownerType->isExistentialType() && reqType->isEqual(ownerType)) {
```

Comment:
Seems like you need to check if `selfType` is actually `Self` or statically refers to `Self`. Right now, this is assuming any time you see the protocol in the right side of the requirement, it's redundant.

Reply:
Hmm, I am not actually using `selfType` at all (except when passing it to `Diags.diagnose`). It can be removed I guess? Or do you mean I need to check `reqType`?

Reply:
Hmm, I actually tried doing that initially, but it fails to work, at least on the example below, because `selfType` is generic:

```
protocol Foo {}
extension Foo where Self: Foo {}
```

In this case, `selfType` is returned as `(generic_type_param_type depth=0 index=0 decl=test1.(file).Foo extension.Self)`.

Reply:
I think if I do `selfType->isEqual(extDecl->getSelfInterfaceType()` then it works!

Reply:
@harlanhaskins If you can run @swift-ci again then I think the tests should pass now!